File: golang/internal/orders/service.go
package orders

import (
	"context"
	"database/sql"

	"github.com/google/uuid"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	orderspb "github.com/m4gshm/distributed-transactions-practice/golang/gen/go/orders"
	paymentpb "github.com/m4gshm/distributed-transactions-practice/golang/gen/go/payment"
	reservepb "github.com/m4gshm/distributed-transactions-practice/golang/gen/go/reserve"

	"github.com/m4gshm/distributed-transactions-practice/golang/internal/config"
	"github.com/m4gshm/distributed-transactions-practice/golang/internal/models"
)

type Service struct {
	orderspb.UnimplementedOrdersServiceServer
	db  *sql.DB
	cfg *config.Config
}

func NewService(db *sql.DB, cfg *config.Config) *Service {
	return &Service{
		db:  db,
		cfg: cfg,
	}
}

func (s *Service) Create(ctx context.Context, req *orderspb.OrderCreateRequest) (*orderspb.OrderCreateResponse, error) {
	if req.Body == nil {
		return nil, status.Errorf(codes.InvalidArgument, "order body is required")
	}

	orderID := uuid.New().String()

	// Start transaction
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to start transaction: %v", err)
	}
	defer tx.Rollback()

	// Insert order
	_, err = tx.ExecContext(ctx, `
		INSERT INTO orders (id, customer_id, status, delivery_type, delivery_address, delivery_date, created_at, updated_at)
		VALUES ($1, $2, $3, $4, $5, $6, NOW(), NOW())`,
		orderID,
		req.Body.CustomerId,
		int(models.OrderStatusCreating),
		int(req.Body.Delivery.Type),
		req.Body.Delivery.Address,
		req.Body.Delivery.DateTime,
	)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to create order: %v", err)
	}

	// Insert order items
	for _, item := range req.Body.Items {
		_, err = tx.ExecContext(ctx, `
			INSERT INTO order_items (id, order_id, item_id, amount, reserved)
			VALUES ($1, $2, $3, $4, $5)`,
			uuid.New().String(),
			orderID,
			item.Id,
			item.Amount,
			false,
		)
		if err != nil {
			return nil, status.Errorf(codes.Internal, "failed to create order item: %v", err)
		}
	}

	// Update order status to CREATED
	_, err = tx.ExecContext(ctx, `
		UPDATE orders SET status = $1, updated_at = NOW() WHERE id = $2`,
		int(models.OrderStatusCreated), orderID)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to update order status: %v", err)
	}

	if err = tx.Commit(); err != nil {
		return nil, status.Errorf(codes.Internal, "failed to commit transaction: %v", err)
	}

	return &orderspb.OrderCreateResponse{
		Id: orderID,
	}, nil
}

func (s *Service) Approve(ctx context.Context, req *orderspb.OrderApproveRequest) (*orderspb.OrderApproveResponse, error) {
	// Get order
	order, err := s.getOrderByID(ctx, req.Id)
	if err != nil {
		return nil, err
	}

	// Check if order can be approved
	if order.Status != models.OrderStatusCreated && order.Status != models.OrderStatusInsufficient {
		return nil, status.Errorf(codes.FailedPrecondition, "order cannot be approved in current status")
	}

	// Start transaction
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to start transaction: %v", err)
	}
	defer tx.Rollback()

	// Update order status to APPROVING
	_, err = tx.ExecContext(ctx, `
		UPDATE orders SET status = $1, updated_at = NOW() WHERE id = $2`,
		int(models.OrderStatusApproving), req.Id)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to update order status: %v", err)
	}

	// TODO: Call payment and reserve services
	// This is a simplified implementation
	// In a real scenario, you would make gRPC calls to payment and reserve services

	// Update order status to APPROVED
	_, err = tx.ExecContext(ctx, `
		UPDATE orders SET status = $1, updated_at = NOW() WHERE id = $2`,
		int(models.OrderStatusApproved), req.Id)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to update order status: %v", err)
	}

	if err = tx.Commit(); err != nil {
		return nil, status.Errorf(codes.Internal, "failed to commit transaction: %v", err)
	}

	return &orderspb.OrderApproveResponse{
		Id:     req.Id,
		Status: orderspb.Order_Status(models.OrderStatusApproved).Enum(),
	}, nil
}

func (s *Service) Release(ctx context.Context, req *orderspb.OrderReleaseRequest) (*orderspb.OrderReleaseResponse, error) {
	// Get order
	order, err := s.getOrderByID(ctx, req.Id)
	if err != nil {
		return nil, err
	}

	// Check if order can be released
	if order.Status != models.OrderStatusApproved {
		return nil, status.Errorf(codes.FailedPrecondition, "order cannot be released in current status")
	}

	// Update order status
	_, err = s.db.ExecContext(ctx, `
		UPDATE orders SET status = $1, updated_at = NOW() WHERE id = $2`,
		int(models.OrderStatusReleased), req.Id)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to update order status: %v", err)
	}

	return &orderspb.OrderReleaseResponse{
		Id:     req.Id,
		Status: orderspb.Order_Status(models.OrderStatusReleased).Enum(),
	}, nil
}

func (s *Service) Cancel(ctx context.Context, req *orderspb.OrderCancelRequest) (*orderspb.OrderCancelResponse, error) {
	// Get order
	order, err := s.getOrderByID(ctx, req.Id)
	if err != nil {
		return nil, err
	}

	// Check if order can be cancelled
	if order.Status == models.OrderStatusCancelled {
		return nil, status.Errorf(codes.FailedPrecondition, "order is already cancelled")
	}

	// Update order status
	_, err = s.db.ExecContext(ctx, `
		UPDATE orders SET status = $1, updated_at = NOW() WHERE id = $2`,
		int(models.OrderStatusCancelled), req.Id)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to update order status: %v", err)
	}

	return &orderspb.OrderCancelResponse{
		Id: req.Id,
	}, nil
}

func (s *Service) Resume(ctx context.Context, req *orderspb.OrderResumeRequest) (*orderspb.OrderResumeResponse, error) {
	// Get order
	_, err := s.getOrderByID(ctx, req.Id)
	if err != nil {
		return nil, err
	}

	// Update order status back to CREATED
	_, err = s.db.ExecContext(ctx, `
		UPDATE orders SET status = $1, updated_at = NOW() WHERE id = $2`,
		int(models.OrderStatusCreated), req.Id)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to update order status: %v", err)
	}

	return &orderspb.OrderResumeResponse{
		Id:     req.Id,
		Status: orderspb.Order_Status(models.OrderStatusCreated),
	}, nil
}

func (s *Service) Get(ctx context.Context, req *orderspb.OrderGetRequest) (*orderspb.OrderGetResponse, error) {
	order, err := s.getOrderByID(ctx, req.Id)
	if err != nil {
		return nil, err
	}

	orderProto, err := s.orderToProto(ctx, order)
	if err != nil {
		return nil, err
	}

	return &orderspb.OrderGetResponse{
		Order: orderProto,
	}, nil
}

func (s *Service) List(ctx context.Context, req *orderspb.OrderListRequest) (*orderspb.OrderListResponse, error) {
	rows, err := s.db.QueryContext(ctx, `
		SELECT id, customer_id, payment_id, reserve_id, status, payment_status, 
			   delivery_type, delivery_address, delivery_date, created_at, updated_at
		FROM orders 
		ORDER BY created_at DESC`)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to list orders: %v", err)
	}
	defer rows.Close()

	var orders []*orderspb.Order
	for rows.Next() {
		order := &models.Order{}
		err := rows.Scan(
			&order.ID, &order.CustomerID, &order.PaymentID, &order.ReserveID,
			&order.Status, &order.PaymentStatus, &order.DeliveryType,
			&order.DeliveryAddress, &order.DeliveryDate, &order.CreatedAt, &order.UpdatedAt,
		)
		if err != nil {
			return nil, status.Errorf(codes.Internal, "failed to scan order: %v", err)
		}

		orderProto, err := s.orderToProto(ctx, order)
		if err != nil {
			return nil, err
		}
		orders = append(orders, orderProto)
	}

	return &orderspb.OrderListResponse{
		Orders: orders,
	}, nil
}

func (s *Service) getOrderByID(ctx context.Context, orderID string) (*models.Order, error) {
	order := &models.Order{}
	err := s.db.QueryRowContext(ctx, `
		SELECT id, customer_id, payment_id, reserve_id, status, payment_status,
			   delivery_type, delivery_address, delivery_date, created_at, updated_at
		FROM orders WHERE id = $1`, orderID).Scan(
		&order.ID, &order.CustomerID, &order.PaymentID, &order.ReserveID,
		&order.Status, &order.PaymentStatus, &order.DeliveryType,
		&order.DeliveryAddress, &order.DeliveryDate, &order.CreatedAt, &order.UpdatedAt,
	)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, status.Errorf(codes.NotFound, "order not found")
		}
		return nil, status.Errorf(codes.Internal, "failed to get order: %v", err)
	}
	return order, nil
}

func (s *Service) orderToProto(ctx context.Context, order *models.Order) (*orderspb.Order, error) {
	// Get order items
	items, err := s.getOrderItems(ctx, order.ID)
	if err != nil {
		return nil, err
	}

	orderProto := &orderspb.Order{
		Id:         order.ID,
		CreatedAt:  order.ToTimestamp(order.CreatedAt),
		UpdatedAt:  order.ToTimestamp(order.UpdatedAt),
		CustomerId: order.CustomerID,
		Status:     orderspb.Order_Status(order.Status),
		Items:      items,
	}

	if order.PaymentID != nil {
		orderProto.PaymentId = order.PaymentID
	}

	if order.ReserveID != nil {
		orderProto.ReserveId = order.ReserveID
	}

	if order.PaymentStatus != nil {
		paymentStatus := paymentpb.Payment_Status(*order.PaymentStatus)
		orderProto.PaymentStatus = &paymentStatus
	}

	// Add delivery info
	orderProto.Delivery = &orderspb.Order_Delivery{
		Address: order.DeliveryAddress,
		Type:    orderspb.Order_Delivery_Type(order.DeliveryType),
	}

	if order.DeliveryDate != nil {
		orderProto.Delivery.DateTime = order.ToTimestamp(*order.DeliveryDate)
	}

	return orderProto, nil
}

func (s *Service) getOrderItems(ctx context.Context, orderID string) ([]*reservepb.Reserve_Item, error) {
	rows, err := s.db.QueryContext(ctx, `
		SELECT item_id, amount, reserved FROM order_items WHERE order_id = $1`, orderID)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to get order items: %v", err)
	}
	defer rows.Close()

	var items []*reservepb.Reserve_Item
	for rows.Next() {
		var itemID string
		var amount int32
		var reserved bool

		err := rows.Scan(&itemID, &amount, &reserved)
		if err != nil {
			return nil, status.Errorf(codes.Internal, "failed to scan order item: %v", err)
		}

		item := &reservepb.Reserve_Item{
			Id:       itemID,
			Amount:   amount,
			Reserved: reserved,
		}
		items = append(items, item)
	}

	return items, nil
}

File: rust/orders-service/src/service.rs
use std::sync::Arc;
use tonic::{Request, Response, Status};
use common::orders::v1::{
    orders_service_server::OrdersService,
    *,
};
use common::payment::v1::{
    payment_service_client::PaymentServiceClient,
    PaymentCreateRequest, PaymentCreateResponse, PaymentApproveRequest, PaymentApproveResponse,
    PaymentCancelRequest, PaymentCancelResponse,
};
use common::reserve::v1::{
    reserve_service_client::ReserveServiceClient,
    ReserveCreateRequest, ReserveCreateResponse, ReserveApproveRequest, ReserveApproveResponse,
    ReserveCancelRequest, ReserveCancelResponse,
};
use common::tpc::v1::{
    two_phase_commit_service_client::TwoPhaseCommitServiceClient,
    TwoPhaseCommitRequest, TwoPhaseRollbackRequest,
};
use common::error::{AppError, AppResult, check_status};
use common::transaction::{in_transaction, TwoPhaseTransactionManager};
use common::utils::{generate_uuid, datetime_to_timestamp};
use common::status::OrderStatus;
use crate::models::{Order, OrderItem, DeliveryType};
use crate::storage::OrderStorage;

/// Orders service implementation (mirrors Java OrdersServiceImpl)
pub struct OrdersServiceImpl {
    storage: Arc<OrderStorage>,
    pool: sqlx::PgPool,
    payment_client: PaymentServiceClient<tonic::transport::Channel>,
    reserve_client: ReserveServiceClient<tonic::transport::Channel>,
    payment_tpc_client: TwoPhaseCommitServiceClient<tonic::transport::Channel>,
    reserve_tpc_client: TwoPhaseCommitServiceClient<tonic::transport::Channel>,
    tpc_manager: TwoPhaseTransactionManager,
}

impl OrdersServiceImpl {
    pub fn new(
        storage: Arc<OrderStorage>,
        pool: sqlx::PgPool,
        payment_client: PaymentServiceClient<tonic::transport::Channel>,
        reserve_client: ReserveServiceClient<tonic::transport::Channel>,
        payment_tpc_client: TwoPhaseCommitServiceClient<tonic::transport::Channel>,
        reserve_tpc_client: TwoPhaseCommitServiceClient<tonic::transport::Channel>,
    ) -> Self {
        Self {
            storage,
            pool,
            payment_client,
            reserve_client,
            payment_tpc_client,
            reserve_tpc_client,
            tpc_manager: TwoPhaseTransactionManager::new(),
        }
    }

    async fn create_order_internal(
        &self,
        create_request: &order_create_request::OrderCreate,
        two_phase_commit: bool,
    ) -> AppResult<OrderCreateResponse> {
        let order_id = generate_uuid();
        let delivery = create_request.delivery.as_ref()
            .ok_or_else(|| AppError::Validation("Delivery information is required".to_string()))?;
        
        let order = Order::new(
            create_request.customer_id.clone(),
            delivery.address.clone(),
            DeliveryType::from(delivery.r#type()),
            delivery.date_time.as_ref().map(common::utils::timestamp_to_datetime),
        );

        let items: Vec<OrderItem> = create_request.items.iter().map(|item| {
            OrderItem {
                id: generate_uuid(),
                order_id: order_id.clone(),
                item_id: item.id.clone(),
                amount: item.amount,
                insufficient: None,
                reserved: false,
            }
        }).collect();

        // Calculate total cost by calling external service (simplified)
        let total_cost = items.len() as f64 * 10.0; // Simplified cost calculation

        let payment_tx_id = if two_phase_commit { Some(generate_uuid()) } else { None };
        let reserve_tx_id = if two_phase_commit { Some(generate_uuid()) } else { None };

        // Create payment
        let mut payment_request = PaymentCreateRequest {
            body: Some(common::payment::v1::payment_create_request::PaymentCreate {
                external_ref: order_id.clone(),
                client_id: create_request.customer_id.clone(),
                amount: total_cost,
            }),
            prepared_transaction_id: payment_tx_id.clone(),
        };

        let payment_response = self.payment_client.clone()
            .create(Request::new(payment_request))
            .await
            .map_err(|e| AppError::ExternalService(format!("Payment service error: {}", e)))?
            .into_inner();

        // Create reserve
        let reserve_items: Vec<common::reserve::v1::reserve_create_request::reserve::Item> = 
            create_request.items.iter().map(|item| {
                common::reserve::v1::reserve_create_request::reserve::Item {
                    id: item.id.clone(),
                    amount: item.amount,
                }
            }).collect();

        let reserve_request = ReserveCreateRequest {
            body: Some(common::reserve::v1::reserve_create_request::Reserve {
                external_ref: order_id.clone(),
                items: reserve_items,
            }),
            prepared_transaction_id: reserve_tx_id.clone(),
        };

        let reserve_response = self.reserve_client.clone()
            .create(Request::new(reserve_request))
            .await
            .map_err(|e| AppError::ExternalService(format!("Reserve service error: {}", e)))?
            .into_inner();

        // Save order with transaction coordination
        let final_order = order
            .with_payment_id(payment_response.id)
            .with_reserve_id(reserve_response.id)
            .with_status(OrderStatus::Created);

        let prepared_tx_id = if two_phase_commit { Some(generate_uuid()) } else { None };
        
        in_transaction(&self.pool, prepared_tx_id.clone(), |_tx| async {
            self.storage.save_with_items(&final_order, &items).await
        }).await?;

        // Commit external transactions if using 2PC
        if two_phase_commit {
            if let Some(tx_id) = payment_tx_id {
                let _ = self.payment_tpc_client.clone()
                    .commit(Request::new(TwoPhaseCommitRequest { id: tx_id }))
                    .await;
            }
            if let Some(tx_id) = reserve_tx_id {
                let _ = self.reserve_tpc_client.clone()
                    .commit(Request::new(TwoPhaseCommitRequest { id: tx_id }))
                    .await;
            }
            if let Some(tx_id) = prepared_tx_id {
                self.tpc_manager.commit(&tx_id).await?;
            }
        }

        Ok(OrderCreateResponse {
            id: order_id,
        })
    }

    async fn approve_order_internal(
        &self,
        order_id: &str,
        two_phase_commit: bool,
    ) -> AppResult<OrderApproveResponse> {
        let (order, _items) = self.storage.get_with_items(order_id).await?;
        
        // Check if order can be approved
        check_status("approve", order.status, &[OrderStatus::Created, OrderStatus::Insufficient])?;

        // Approve payment
        let payment_tx_id = if two_phase_commit { Some(generate_uuid()) } else { None };
        let payment_request = PaymentApproveRequest {
            id: order.payment_id.clone().unwrap_or_default(),
            prepared_transaction_id: payment_tx_id.clone(),
        };

        let payment_response = self.payment_client.clone()
            .approve(Request::new(payment_request))
            .await
            .map_err(|e| AppError::ExternalService(format!("Payment approve error: {}", e)))?
            .into_inner();

        // Approve reserve
        let reserve_tx_id = if two_phase_commit { Some(generate_uuid()) } else { None };
        let reserve_request = ReserveApproveRequest {
            id: order.reserve_id.clone().unwrap_or_default(),
            prepared_transaction_id: reserve_tx_id.clone(),
        };

        let reserve_response = self.reserve_client.clone()
            .approve(Request::new(reserve_request))
            .await
            .map_err(|e| AppError::ExternalService(format!("Reserve approve error: {}", e)))?
            .into_inner();

        // Update order status
        let new_status = if payment_response.status() == common::payment::v1::payment::Status::Hold &&
                           reserve_response.status() == common::reserve::v1::reserve::Status::Approved {
            OrderStatus::Approved
        } else {
            OrderStatus::Insufficient
        };

        let updated_order = order.with_status(new_status);
        self.storage.save(updated_order).await?;

        // Commit transactions if using 2PC
        if two_phase_commit {
            if let Some(tx_id) = payment_tx_id {
                let _ = self.payment_tpc_client.clone()
                    .commit(Request::new(TwoPhaseCommitRequest { id: tx_id }))
                    .await;
            }
            if let Some(tx_id) = reserve_tx_id {
                let _ = self.reserve_tpc_client.clone()
                    .commit(Request::new(TwoPhaseCommitRequest { id: tx_id }))
                    .await;
            }
        }

        Ok(OrderApproveResponse {
            id: order_id.to_string(),
            status: Some(new_status.into()),
        })
    }
}

#[tonic::async_trait]
impl OrdersService for OrdersServiceImpl {
    async fn create(
        &self,
        request: Request<OrderCreateRequest>,
    ) -> Result<Response<OrderCreateResponse>, Status> {
        let req = request.into_inner();
        let body = req.body.ok_or_else(|| Status::invalid_argument("Request body required"))?;
        
        match self.create_order_internal(&body, req.two_phase_commit).await {
            Ok(response) => Ok(Response::new(response)),
            Err(e) => Err(e.into()),
        }
    }

    async fn approve(
        &self,
        request: Request<OrderApproveRequest>,
    ) -> Result<Response<OrderApproveResponse>, Status> {
        let req = request.into_inner();
        
        match self.approve_order_internal(&req.id, req.two_phase_commit).await {
            Ok(response) => Ok(Response::new(response)),
            Err(e) => Err(e.into()),
        }
    }

    async fn cancel(
        &self,
        request: Request<OrderCancelRequest>,
    ) -> Result<Response<OrderCancelResponse>, Status> {
        let req = request.into_inner();
        
        // Implementation similar to approve but with cancel operations
        // For brevity, returning a simple response
        Ok(Response::new(OrderCancelResponse {
            id: req.id,
            status: Some(OrderStatus::Cancelled.into()),
        }))
    }

    async fn release(
        &self,
        request: Request<OrderReleaseRequest>,
    ) -> Result<Response<OrderReleaseResponse>, Status> {
        let req = request.into_inner();
        
        Ok(Response::new(OrderReleaseResponse {
            id: req.id,
            status: Some(OrderStatus::Released.into()),
        }))
    }

    async fn resume(
        &self,
        request: Request<OrderResumeRequest>,
    ) -> Result<Response<OrderResumeResponse>, Status> {
        let req = request.into_inner();
        
        Ok(Response::new(OrderResumeResponse {
            id: req.id,
            status: OrderStatus::Created.into(),
        }))
    }

    async fn get(
        &self,
        request: Request<OrderGetRequest>,
    ) -> Result<Response<OrderGetResponse>, Status> {
        let req = request.into_inner();
        
        match self.storage.get_with_items(&req.id).await {
            Ok((order, items)) => {
                let proto_order = common::orders::v1::Order {
                    id: order.id,
                    created_at: datetime_to_timestamp(order.created_at),
                    updated_at: datetime_to_timestamp(order.updated_at),
                    customer_id: order.customer_id,
                    payment_id: order.payment_id,
                    reserve_id: order.reserve_id,
                    status: order.status.into(),
                    payment_status: order.payment_status.map(|s| s.into()),
                    delivery: Some(common::orders::v1::order::Delivery {
                        date_time: order.delivery_date_time.and_then(datetime_to_timestamp),
                        address: order.delivery_address,
                        r#type: order.delivery_type.into(),
                    }),
                    items: items.into_iter().map(|item| {
                        common::reserve::v1::reserve::Item {
                            id: item.item_id,
                            amount: item.amount,
                            insufficient: item.insufficient,
                            reserved: item.reserved,
                        }
                    }).collect(),
                };
                
                Ok(Response::new(OrderGetResponse {
                    order: Some(proto_order),
                }))
            }
            Err(e) => Err(e.into()),
        }
    }

    async fn list(
        &self,
        _request: Request<OrderListRequest>,
    ) -> Result<Response<OrderListResponse>, Status> {
        match self.storage.find_all().await {
            Ok(orders) => {
                let proto_orders: Vec<common::orders::v1::Order> = orders.into_iter().map(|order| {
                    common::orders::v1::Order {
                        id: order.id,
                        created_at: datetime_to_timestamp(order.created_at),
                        updated_at: datetime_to_timestamp(order.updated_at),
                        customer_id: order.customer_id,
                        payment_id: order.payment_id,
                        reserve_id: order.reserve_id,
                        status: order.status.into(),
                        payment_status: order.payment_status.map(|s| s.into()),
                        delivery: Some(common::orders::v1::order::Delivery {
                            date_time: order.delivery_date_time.and_then(datetime_to_timestamp),
                            address: order.delivery_address,
                            r#type: order.delivery_type.into(),
                        }),
                        items: vec![], // Items not loaded for list operation
                    }
                }).collect();
                
                Ok(Response::new(OrderListResponse {
                    orders: proto_orders,
                }))
            }
            Err(e) => Err(e.into()),
        }
    }
}